Add Codesgiving tip of the week blog post

d33bs opened the pull request:
This PR adds a blog post covering something nicknamed "Codesgiving" and an open-source contribution walkthrough. I tried to make this a "map/guide" on how the open-source contribution process works, including several general software strategies as well, while channeling the spirit of the upcoming US holiday.

I plan to follow up with an "applied" open-source contribution post later, exploring a real-world open-source contribution and how it worked (recognizing the process isn't always so clear / precise).

Thanks in advance for your thoughts and feedback on this! And, importantly, wishing you all a great upcoming Thanksgiving!

Note: please take your time, no pressure to review before the holiday.

CC @jenna-tomkinson

falquaddoomi commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+  var spec = {
+    "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
+    "description": "Suggested Open-source Time Allocations",
+    "width": 300,
```
I've noticed that the legend text gets ellipsized at this width. Perhaps the entire figure or just the legend could be made wider?

It's not a blocker since you can still hover over the corresponding slice to see a tooltip with the full text (which, I might add, is a cool feature), but I thought I'd mention it anyway.

vincerubinetti replied:
Related, might want to add this for smaller screen widths:

```html
<style>
canvas {
  max-width: 100%;
  height: unset;
}
</style>
```

Or if vega allows `width` to be "auto" (full width), I think that'd be fine. Or a max-width of 300px.

d33bs replied:
Thanks - great points and suggestions! I'll see about improving this.

falquaddoomi commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+Notice the emphasis on planning and testing and keep these in mind as you progress (the actual programming time can be small if adequate time has been spent on planning).
+Notably, the original time fractions are modified here with the final quarter of the time spent suggested as code review, revisions, and post-actions.
+Planning for the time expense of the added code review and related elements assists with keeping a learning mindset throughout the process (instead of feeling like the review is a "tack-on" or "optional / supplementary").
+A good motto to keep in mind throughout this process is [_Festina lente_](https://en.wikipedia.org/wiki/Festina_lente), or __"Make haste, slowly."__ (take care to move thoughtfully and as slow as necessary to do thing correctly the first time).
```
```suggestion
A good motto to keep in mind throughout this process is [_Festina lente_](https://en.wikipedia.org/wiki/Festina_lente), or __"Make haste, slowly."__ (take care to move thoughtfully and as slowly as necessary to do things correctly the first time).
```

d33bs replied:
Thanks, reads better this way!

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+
+<!-- excerpt start -->
+[Thanksgiving](https://en.wikipedia.org/wiki/Thanksgiving) is a holiday practiced in many countries which focuses on gratitude for good harvests of the preceding year.
+In the United States, we celebrate Thanksgiving on November 23rd each year often by eating meals we create together with others.
```
I think it's on the fourth Thursday each year so not always the 23rd?

d33bs replied:
Thanks, oops, going to correct this!

falquaddoomi commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+One template for reporting issues is through a "user story".
+A user story typically comes in the form: `As a < type of user >, I want < some goal > so that < some reason >.` ([Mountain Goat Software: User Stories](https://www.mountaingoatsoftware.com/agile/user-stories)).
+Alongside the story, it can help to add in a snippet of code which exemplifies a problem, new functionality, or a potential adjacent / similar solution.
+As a general principle, be as __be as specific as you can without going overboard__.
```
```suggestion
As a general principle, __be as specific as you can without going overboard__.
```

d33bs replied:
Thanks!

falquaddoomi commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+Alongside the story, it can help to add in a snippet of code which exemplifies a problem, new functionality, or a potential adjacent / similar solution.
+As a general principle, be as __be as specific as you can without going overboard__.
+Include things like programming language version, operating system, and other system dependencies that might be related.
+Once you have a good written description of the need be sure to submit it where it can be seen by the relevant development community.
```
```suggestion
Once you have a good written description of the need, be sure to submit it where it can be seen by the relevant development community.
```

d33bs replied:
Great catch!

falquaddoomi commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+
+#### Bug or Feature Verification with Test-driven Development
+
+> 1. Add or modify a test which checks for a bug fix or feature addition
+> 2. Run all tests (expecting the newly added test content to fail)
+> 3. Write a simple version of code which allows the tests to succeed
+> 4. Verify that all tests now pass
+> 5. Return to step 3, refactoring the code as needed
+
```
Are these steps supposed to be below the line further down that says "Test-driven development entails the following pattern:"?

d33bs replied:
Whoops, yes, thanks! I tried to make these steps a sort of "graphic" ahead of the text, but they lost the improved flow you mention. I'll work this in somehow with an additional commit for this PR.

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+Planning for the time expense of the added code review and related elements assists with keeping a learning mindset throughout the process (instead of feeling like the review is a "tack-on" or "optional / supplementary").
+A good motto to keep in mind throughout this process is [_Festina lente_](https://en.wikipedia.org/wiki/Festina_lente), or __"Make haste, slowly."__ (take care to move thoughtfully and as slow as necessary to do thing correctly the first time).
+
+1. 1/3 planning (~33%)
```
Perhaps move this to right below the chart, since it's directly related? Good on you for putting it in text, as these data visualizations are seldom accessible.

d33bs replied:
Great idea, thanks! I'll move this as you mentioned.

falquaddoomi reviewed (commented):
I left a few comments about possible typos, but aside from that, this is a pretty meaty guide to open-source contributions. Nice work!

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+
+Be sure to check whether the bug or feature has already been reported somewhere!
+In a way, this is a practice of ["Don't repeat yourself" (DRY)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) where we attempt to avoid repeating the same block of code (in this case, the "code" can be understood as natural language).
+For example, you can look on GitHub Issues with a search query matching the rough idea of what you're thinking about.
```
Very minor point, but maybe mention GitHub Discussions too, because I feel like people forget about this usually.

d33bs replied:
Such a good point, thanks! I'll find a way to work this in. I wonder if there's a way to mention the new search bar functionality to this effect (unsure if it highlights discussions alongside issues).

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+In a way, this is a practice of ["Don't repeat yourself" (DRY)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) where we attempt to avoid repeating the same block of code (in this case, the "code" can be understood as natural language).
+For example, you can look on GitHub Issues with a search query matching the rough idea of what you're thinking about.
+If it has been reported already, take a look to see if someone has made a code contribution related to the work already.
+If there aren't yet any code contributions and it doesn't look like anyone is working on one, consider volunteering to take a further look into the solution and be sure to acknowledge the existing discussions.
```
Another very minor point, but I find that it helps to acknowledge the design difficulties and constraints of the author/maintainer, e.g. "i can see how creating a flexible and simple API would be hard for this, but maybe you could do something like this".

d33bs replied:
Love this, showing sympathy or empathy can go a long way. Thanks - I'll find a way to add this in.

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+When your code and new test(s) are in a good spot it's time to ask for a code review.
+Be sure to follow the open-source [`CONTRIBUTING.md` document](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors) documentation related to the project; each one can vary.
+When working on GitHub-based projects, you'll need to open a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) on the correct branch (usually upstream `main`).
+If you used a GitHub issue to help report the issue, mention the issue in the pull request description using the `#<issue number` reference to help link the work to the reported need (this will cause the pull request to show up within the issue).
```
`#issue number`?

d33bs replied:
Thanks, much better!

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+
+Once the code has been accepted by the reviewers and through potential automated testing suite(s) the content is ready to be merged.
+Oftentimes this work is completed by core maintainers of the project.
+After the code is merged, be sure to take time to clean up your workspace by deleting your development branch and syncing with the upstream repository (optional).
```
"it's usually a good idea to clean up..." rather than "(optional)"?

d33bs replied:
Thanks, that sounds much better - I'll work this in!

vincerubinetti commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+When working on GitHub-based projects, you'll need to open a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) on the correct branch (usually upstream `main`).
+If you used a GitHub issue to help report the issue, mention the issue in the pull request description using the `#<issue number` reference to help link the work to the reported need (this will cause the pull request to show up within the issue).
+
+Note: It might feel tempting to perfect the code.
```
I feel like this point is said several times in article. Maybe consider reviewing if any of the instances can be removed. Though it is certainly a good thing to emphasize multiple times, and keep in mind at all times during the process.

vincerubinetti reviewed (approved):
Very nicely written. I made some comments, none are essential for merge.

d33bs commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+> \- Albert Camus
+
+These open-source projects have internal costs which are sometimes invisible to those who consume them.
+Every software project has an implied level of [software gardening](https://bssw.io/blog_posts/long-term-software-gardening-strategies-for-cultivating-scientific-development-ecosystems) time costs involved to impede decay,  practice continuous improvements and evolve the work.
```
Great catch, thanks!

d33bs commented on _posts/2023-11-15-Codesgiving-Part-1.md:
```diff
@@ -0,0 +1,289 @@
+Prepare yourself by intentionally opening your mind to input from others, even if you're sure you're absolutely "right".
+
+### Setting Software Scheduling Expectations
+
```
Thanks, will apply this change in a further commit!

d33bs commented:
Thank you @vincerubinetti, @falquaddoomi, and @jenna-tomkinson for your reviews! I've updated the content based on what you mentioned and added a few visual enhancements along the way. Going to merge this in.
